srml/authorship/src/lib.rs

```diff
@@ -267,55 +355,80 @@ impl<T: Trait> Module<T> {
 			(min, now)
 		};
 
-		let mut uncles = <Self as Store>::Uncles::get();
-		uncles.push(UncleEntryItem::InclusionHeight(now));
+		let hash = uncle.hash();
 
-		let mut acc: <T::FilterUncle as FilterUncle<_, _>>::Accumulator = Default::default();
+		if uncle.number() < &One::one() {
+			return Err("uncle is genesis");
+		}
 
-		for uncle in new_uncles {
-			let hash = uncle.hash();
+		if uncle.number() > &maximum_height {
+			return Err("uncle is too high in chain");
+		}
 
-			if uncle.number() < &One::one() {
-				return Err("uncle is genesis");
+		{
+			let parent_number = uncle.number().clone() - One::one();
+			let parent_hash = <system::Module<T>>::block_hash(&parent_number);
+			if &parent_hash != uncle.parent_hash() {
+				return Err("uncle parent not in chain");
 			}
+		}
 
-			if uncle.number() > &maximum_height {
-				return Err("uncles too high in chain");
-			}
+		if uncle.number() < &minimum_height {
+			return Err("uncle not recent enough to be included");
+		}
 
-			{
-				let parent_number = uncle.number().clone() - One::one();
-				let parent_hash = <system::Module<T>>::block_hash(&parent_number);
-				if &parent_hash != uncle.parent_hash() {
-					return Err("uncle parent not in chain");
-				}
-			}
+		let duplicate = existing_uncles.into_iter().find(|h| **h == hash).is_some();
+		let in_chain = <system::Module<T>>::block_hash(uncle.number()) == hash;
 
-			if uncle.number() < &minimum_height {
-				return Err("uncle not recent enough to be included");
-			}
+		if duplicate || in_chain {
+			return Err("uncle already included")
+		}
 
-			let duplicate = uncles.iter().find(|entry| match entry {
-				UncleEntryItem::InclusionHeight(_) => false,
-				UncleEntryItem::Uncle(h, _) => h == &hash,
-			}).is_some();
+		// check uncle validity.
+		T::FilterUncle::filter_uncle(&uncle, accumulator)
+	}
+}
 
-			let in_chain = <system::Module<T>>::block_hash(uncle.number()) == hash;
+impl<T: Trait> ProvideInherent for Module<T> {
+	type Call = Call<T>;
+	type Error = MakeFatalError<()>;
+	const INHERENT_IDENTIFIER: InherentIdentifier = INHERENT_IDENTIFIER;
 
-			if duplicate || in_chain { return Err("uncle already included") }
+	fn create_inherent(data: &InherentData) -> Option<Self::Call> {
+		let uncles = data.uncles().expect("Gets and decodes final number inherent data");
```

Comment:
panicking here is wrong

Reply:
Panic happens when the data has invalid encoding, not when it is missing.

Reply:
yes. still wrong IMO. what proves that the data will have correct encoding? 

you could maybe make a case that the provider provides data that's correct, but it's a bit of a far-reaching assumption, since it can be broken by a change in provider implementation later.